### DISCUSSION TITLE
Progress Update

### DISCUSSION THREAD

**Discussion by paulpv** [DistroAV/DistroAV#840] [General]
I do have a real job and family life that take up what feels like 99% of my time, but I promise y'all that I am making progress on this, although I am certain it is not as fast as any of us would like.

Good news! **I got Audio working on MacOS M1/ARM64 OBS + obs-ndi!!!**

I have done a lot of researching in the past few days on github/reddit/facebook/stackoverflow/etc about MacOS obs-ndi problems for the past 2+ years (that existed even before the latest OBS 28 compatibility issues), including M1/M2/aarm64 support.  
I see a lot of common issues and hacks of manual installs and I have a suspicion about **some of** what is going wrong.

The problem was that it was loading the wrong `libndi`.
When I launch OBS **with the existing obs-ndi** and look at the logs I see:
```
...
20:05:50.923: [obs-ndi] hello ! (version 4.10.1)
20:05:50.923: [obs-ndi] Trying ''
20:05:50.925: [obs-ndi] Trying '/usr/lib'
20:05:50.925: [obs-ndi] Trying '/usr/local/lib'
20:05:50.925: [obs-ndi] Found NDI library at '/usr/local/lib/libndi.4.dylib'
20:05:50.931: [obs-ndi] NDI runtime loaded successfully
20:05:50.931: [obs-ndi] NDI library initialized successfully (NDI SDK APPLE 11:50:41 Apr  1 2020 4.5.1)
...
```
That is clearly loading NDI 4.5.1 even though I was certain I had `NDI Tools 5.5.3` installed.
Yesterday I did a fresh install of the latest `NDI 5.5.3 Tools for MacOS` and noticed:
```
paul.peavyhouse@macbook ~ % ls -la /usr/local/lib
drwxr-xr-x  5 root  wheel       160 Sep 14 16:06 .
drwxr-xr-x  5 root  wheel       160 Feb 17 15:09 ..
-rwxr-xr-x  1 root  wheel   4132576 Apr  1  2020 libndi.4.dylib
-rwxr-xr-x  1 root  wheel  28247296 Jul 29  2022 libndi.dylib
-rw-r--r--  1 root  wheel      7610 Apr  1  2020 libndi_licenses.txt
```
Notice that there are two separate `libndi`s, and neither is a symlink to the other.
There is a discrete v4 `libndi.4.dylib` and v5 `libndi.dylib` binary.

Today I installed `NDI 5.5.3 SDK for Apple` and noticed:
```
paul.peavyhouse@macbook ~ % ls -la /Library/NDI\ SDK\ for\ Apple/lib/macOS/      
drwxr-xr-x  4 root  wheel       128 Feb 18 17:02 .
drwxr-xr-x  5 root  wheel       160 Aug  9  2022 ..
-rwxr-xr-x  1 root  wheel  28396160 Feb  2 08:00 libndi.dylib
-rw-r--r--  1 root  wheel     12577 Feb  2 08:00 libndi_licenses.txt
```
Notice that the two `libndi.dylib` files are different, and there is no `libndi.4.dylib`.

Then I deleted `sudo rm -r /usr/local/lib/libndi*` and used the `NDILIB_REDIST_URL    "http://new.tk/NDIRedistV5Apple"` mentioned in the code:
https://github.com/paulpv/obs-ndi/blob/develop/lib/ndi/Processing.NDI.Lib.h#L64
Now I have:
```
paul.peavyhouse@macbook Downloads % ls -la /usr/local/lib         
drwxr-xr-x  4 root  wheel       128 Feb 18 21:50 .
drwxr-xr-x  5 root  wheel       160 Feb 18 21:50 ..
-rwxr-xr-x  1 root  wheel  28396160 Feb  2 08:01 libndi.dylib
-rw-r--r--  1 root  wheel     12577 Feb  2 08:01 libndi_licenses.txt
```

And now the existing obs-ndi plugin cannot find `libndi.4.dylib`.

The good news is the patch to [my fork](https://github.com/paulpv/obs-ndi/tree/develop) of this obs-ndi plugin [already has a fix for this](https://github.com/paulpv/obs-ndi/commit/ea694b166a3bf91fb2e9afa09f435e3ac91410d7).

So, my fork now loads the correct `libndi` and audio is now output.

The bad news is that the video output is still less than 1 FPS. :/
**So, I am still looking in to this and not yet ready to submit a PR for merge and release of the next official version...**
...but I am closer!

I am trying to to release this until all of the following builds have basic working functionality:
* `Windows x64`: Known blocking issues: **None**
* `Linux x64`: Known blocking issues: **None**
* `Linux ARM64`: Known blocking issues: **None**
* `MacOS Mx/ARM64`: Known blocking issues: **Video output less than 1 FPS**

Working `Windows ARM64`, `MacOS x64`, and `RPi ARM64` would be awesome, but are stretch goals that will not block releasing the next official version.  
Feel free to flame me if you disagree.

**Comment by paulpv** [comment]
Update: I got MacOS video output to have expected FPS by replacing `send_send_video_v2` with `send_send_video_async_v2`!

I guess the sync call really was too much overhead!

This adds the problem of needing to ping-pong alternate between two video frame buffers, like @tt2468 implemented in his `rewrite` branch, otherwise video sheering may happen (although I am not seeing it) or the app may even crash if something tries to free the video frame buffer (also something I have not seen yet).

I can do the same thing, but it is 1am my time, so I will be pausing for now and pick it up again soon, hopefully tomorrow.
I am going to leave an NDI output stream running until I pick it back up again, to check for leaks/crash.

I don't have Presidents Day off, so no long weekend for me, unfortunately.

**Comment by jbcurler2010** [comment]
Paul, it goes without saying that there are a ton of people happy that you've taken this on. If I had the skills, I would have done it myself because I understand how important it is. Thanks for all your work.
